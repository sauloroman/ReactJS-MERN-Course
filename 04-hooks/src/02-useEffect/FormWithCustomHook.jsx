import { useForm } from "../hooks/useForm"

export const FormWithCustomHook = () => {

  const { onInputChange, onResetForm, username, email } = useForm({ username: '', email: '' })

  return (
    <>
      <h2 className="heading">UseEffect &mdash; FormCustomHook</h2>

      <form action="#" className="form">
        <div className="form__field">
          <label htmlFor="username" className="form__label">Nombre del usuario</label>
          <input 
            data-testid="username-input"
            id="username"
            name="username"
            value={username}
            onChange={ onInputChange }
            className="form__input" 
            placeholder="Ingrese su nombre" 
            type="text" 
          />
        </div>
        <div className="form__field">
          <label htmlFor="email" className="form__label">Correo del usuario</label>
          <input 
            data-testid="email-input"
            id="email"
            name="email"
            value={email}
            onChange={ onInputChange }
            className="form__input" 
            placeholder="Ingrese su correo" 
            type="text" 
          />
        </div>

        <button onClick={ onResetForm } className="form__btn">Limpiar formulario</button>
      </form>
    </>
  )
}
