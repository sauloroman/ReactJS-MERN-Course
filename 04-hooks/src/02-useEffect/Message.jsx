import { useEffect, useState } from "react"

export const Message = () => {
  
  const [coords, setCoords] = useState({ x: 0, y: 0 })

  useEffect(() => {

    const onMouseMove = (event) => {
      const { clientX, clientY } = event
      setCoords({ x: clientX, y: clientY })
    }

    window.addEventListener('mousemove', onMouseMove )

    return () => {
      window.removeEventListener('mousemove', onMouseMove )
    }

  }, [])

  return (
    <p>Coordenadas: <span data-testid="coords">{ JSON.stringify( coords ) }</span></p>
  )
}
