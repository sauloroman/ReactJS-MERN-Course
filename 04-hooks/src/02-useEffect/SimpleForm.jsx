import { useEffect, useState } from "react"
import { Message } from "./Message"

export const SimpleForm = () => {
  
  const [formState, setFormState] = useState({
    username: '',
    email: '',
    age: '',
  })
  
  const { username, email, age } = formState

  const onInputChange = ({ target: { name, value }}) => {
    setFormState({
      ...formState,
      [name]: value
    })
  }

  // useEffect(() => {
  //   console.log('The component is mounted')
  // }, [])

  // useEffect(() => {
  //   console.log('Username has changed')
  // }, [username])

  // useEffect(() => {
  //   console.log('Email has changed')
  // }, [email])

  // useEffect(() => {
  //   console.log('Age has changed')
  // }, [age])

  return (
    <>
      <h1 className="heading">UseEffect &mdash; Formulario Simple</h1>

      <form action="#" className="form">
        <div className="form__field">
          <label htmlFor="username" className="form__label">Nombre del usuario</label>
          <input 
            name="username"
            value={ username }
            onChange={onInputChange}
            type="text" 
            className="form__input" 
            placeholder="Ingrese su nombre"
          />
        </div>
        <div className="form__field">
          <label htmlFor="email" className="form__label">Correo Electr√≥nico</label>
          <input 
            name="email"
            value={ email }
            onChange={onInputChange}
            type="email" 
            className="form__input" 
            placeholder="Ingrese su correo"
          />
        </div>
        <div className="form__field">
          <label htmlFor="age" className="form__label">Edad</label>
          <input 
            name="age"
            value={ age }
            onChange={onInputChange}
            type="number"
            min={1} 
            className="form__input" 
            placeholder="Ingrese su edad"
          />
        </div>
      </form>

      { username === 'Roman' && <Message /> }
    </>
  )
}
