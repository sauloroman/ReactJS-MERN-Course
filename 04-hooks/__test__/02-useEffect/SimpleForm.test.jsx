import { fireEvent, render, screen } from '@testing-library/react'
import { SimpleForm } from '../../src/02-useEffect/SimpleForm'

describe('Tests in SimpleForm', () => {

  test('It must match with the spanshot', () => {
    const { container } = render(<SimpleForm />)
    expect( container ).toMatchSnapshot()
  })

  test('it must change input values', () => {
    render(<SimpleForm />)

    const inputValue = 'Roman Nava'
    const emailValue = 'roman@gmail.com'
    const ageValue = '20'

    const usernameInput = screen.getByTestId('username-input')
    const emailInput = screen.getByTestId('email-input')
    const ageInput = screen.getByTestId('age-input')

    fireEvent.change( usernameInput, { target: { value: inputValue } })
    fireEvent.change( emailInput, { target: { value: emailValue }})
    fireEvent.change( ageInput, { target: { value: ageValue }} )

    expect( usernameInput.value ).toBe(inputValue)
    expect( emailInput.value ).toBe(emailValue)
    expect( ageInput.value ).toBe( ageValue )

  })

})