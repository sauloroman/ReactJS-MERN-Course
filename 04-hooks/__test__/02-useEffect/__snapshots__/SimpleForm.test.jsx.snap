// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in SimpleForm It must match with the spanshot 1`] = `
<div>
  <h1
    class="heading"
  >
    UseEffect — Formulario Simple
  </h1>
  <form
    action="#"
    class="form"
    data-testid="simple-form"
  >
    <div
      class="form__field"
    >
      <label
        class="form__label"
        for="username"
      >
        Nombre del usuario
      </label>
      <input
        class="form__input"
        data-testid="username-input"
        name="username"
        placeholder="Ingrese su nombre"
        type="text"
        value=""
      />
    </div>
    <div
      class="form__field"
    >
      <label
        class="form__label"
        for="email"
      >
        Correo Electrónico
      </label>
      <input
        class="form__input"
        data-testid="email-input"
        name="email"
        placeholder="Ingrese su correo"
        type="email"
        value=""
      />
    </div>
    <div
      class="form__field"
    >
      <label
        class="form__label"
        for="age"
      >
        Edad
      </label>
      <input
        class="form__input"
        data-testid="age-input"
        min="1"
        name="age"
        placeholder="Ingrese su edad"
        type="number"
        value=""
      />
    </div>
  </form>
</div>
`;
